lib/zhr_devs/bakery_integration/command_runner.ex

```diff
@@ -29,31 +38,43 @@ defmodule ZhrDevs.BakeryIntegration.CommandRunner do
 
     Port.monitor(port)
 
-    {:ok, %State{port: port}}
+    {:ok, %State{port: port, on_success: on_success, on_failure: on_failure}}
   end
 
   # This callback handles data incoming from the command's STDOUT
-  def handle_info({port, {:data, text_line}}, %{port: port} = state) do
-    Logger.info("Data: #{inspect(text_line)}")
+  def handle_info({_port, {:data, text_line}}, state) do
     {:noreply, %State{state | latest_output: String.trim(text_line)}}
   end
 
   # This callback tells us when the process exits
   def handle_info({port, {:exit_status, status}}, %{port: port} = state) do
     Logger.info("Port exit: :exit_status: #{status}")
 
-    new_state = %State{state | exit_status: status}
+    error = %{error: :execution_stopped, context: state.latest_output, exit_status: status}
+    :ok = state.on_error.(error)
 
-    {:noreply, new_state}
+    {:stop, :shutdown, state}
   end
 
   def handle_info({:DOWN, _ref, :port, port, :normal}, state) do
     Logger.info("Handled :DOWN message from port: #{inspect(port)}")
-    {:noreply, state}
+
+    case state.on_success.() do
```

Comment:
Ah, I see. DOWN must be the OTP thing that tells the process that it has to finish and we gracefully handle it here.